### PR TITLE
feat(DTFS2-6935): add test coverage for front-end pagination changes for TFM deals

### PR DESCRIPTION
**Pull request by oscar-richardson-softwire**
## Introduction :pencil2:
This PR adds test coverage for the _front-end_ changes that introduce pagination to the TFM deals table. The backend changes already have test coverage as part of https://github.com/UK-Export-Finance/dtfs2/pull/2561

## Resolution :heavy_check_mark:
- added Cypress tasks to insert/delete TFM deals directly into/from the database to enable efficient testing of pagination
- fixed/added E2E, component/unit and API tests to cover the front-end pagination changes in TFM UI
- refactored pagination component slightly to make it more consistent with other components and to allow it to be tested using our existing `componentRenderer`

## Miscellaneous :heavy_plus_sign:
- the pagination component also exists in Portal UI and was missing component tests. I added tests here as well
- manually tested the TFM deals table pagination with 2000 deals (to mimic the number on production). Page load time decreased from ~4 seconds to 6 milliseconds

### CODE REVIEW

File: e2e-tests/support/user-tasks.js [hunk 1]
```diff
@@ -44,6 +47,22 @@ module.exports = {
           },
         );
       },
+
+      async insertManyTfmDeals(numberOfDealsToInsert) {
```
Comment (ttbarnes) [inline e2e-tests/support/user-tasks.js:51]: Could we add some documentation? Little hard to follow, fair amount of things going on in the for loop.

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it('should allow the user to navigate to a specific page of the deals table', () => {
+    pages.dealsPage.pagination.page(2).click();
+
+    cy.url().should('eq', relative('/deals/2'));
+    pages.dealsPage.dealsTableRows().eq(0).contains('10000041');
+    pages.dealsPage.dealsTableRows().eq(19).contains('10000060');
+    pages.dealsPage.dealsTableRows().should('have.length', 20);
+  });
```
Comment (ttbarnes) [inline e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:71]: Nice work, we could create some cypress commands here to be DRY e.g

```js
cy.checkDealsTableRow({ index: 19, dealNumber: 10000060  });
cy.checkDealsTableTotalRows(20);
```

Also I think we should update the row assertions to check the deal number column specifically instead of `.contains('10000001')`, because if the deal number is present in the row, but not in the deal number column, this would still pass

File: portal/component-tests/_macros/pagination.component-test.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+      wrapper.expectLink('[data-cy="Next"]').notToExist();
+    });
+
+    it('should NOT render a \'Last\' link', () => {
+      wrapper.expectElement('[data-cy="Last_listItem"]').notToExist();
+
+      wrapper.expectLink('[data-cy="Last"]').notToExist();
+    });
+  };
```
Comment (ttbarnes) [inline portal/component-tests/_macros/pagination.component-test.js:136]: Nice DRY tests - similar to Alex's comment above, should we these into their own directory e.g `test-assertions`?

Also is it worth moving all the `data-cy` selectors into their own variables, similar to cypress selectors?

File: portal/component-tests/_macros/pagination.component-test.js [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+    shouldNOTRenderFirstAndPreviousLinksAndEllipsis();
+
+    shouldNOTRenderEllipsisAndNextAndLastLinks();
+  });
+
+  describe('when totalPages is more than 1', () => {
+    describe('when currentPage is 0', () => {
+      beforeEach(() => {
+        totalPages = 100;
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply portal/component-tests/_macros/pagination.component-test.js:164]: Agree @AlexBramhill I'd try to test all possible scenarios. i'm not sure on the pagination design, on EXIPs we have I think 8 possible design scenarios, depending on what page a user is on and how many deals/applications there are so we have an E2E test for each of these scenarios. Very easy with some DRY test assertions  => https://github.com/UK-Export-Finance/exip/tree/main-application-no-pdf/e2e-tests/insurance/cypress/e2e/journeys/dashboard/dashboard-pagination

File: trade-finance-manager-ui/server/api.test.js [hunk 5]
```diff
@@ -15,3 +17,131 @@ describe('getFacilities()', () => {
...
+    expect(axios).toHaveBeenCalledTimes(1);
+    expect(axios).toHaveBeenCalledWith({
+      method: 'get',
+      url: `${TFM_API_URL}/v1/deals`,
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+        'x-api-key': TFM_API_KEY,
+      },
+      params: queryParams,
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/api.test.js:118]: Could we create an e.g `baseExpected` for everything but `params`? Seems to be duplicated throughout

File: e2e-tests/support/user-tasks.js (hunk 1 above)
Comment (oscar-richardson-softwire) [reply e2e-tests/support/user-tasks.js:51]: Added docstring

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    pages.dealsPage.dealsTableRows().eq(0).contains('10000041');
+    pages.dealsPage.dealsTableRows().eq(10).contains('10000061');
+    pages.dealsPage.dealsTableRows().should('have.length', 11);
+  });
+
+  it('should redirect to the first page of the deals table when the user sorts the table', () => {
+    cy.visit('/deals/2');
+    pages.dealsPage.dealsTable.headings.ukefDealIdSortButton().click();
+
+    cy.url().should('eq', relative('/deals/0?sortfield=dealSnapshot.ukefDealId&sortorder=ascending'));
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:119]: Added checks for elements

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it('should redirect to the first page of the deals table when the user searches', () => {
+    cy.visit('/deals/2');
+    const searchString = 'Company 1';
+    pages.dealsPage.searchFormInput().type(searchString);
+    pages.dealsPage.searchFormSubmitButton().click();
+
+    cy.url().should('eq', relative('/deals/0?search=Company%201'));
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:128]: Added checks for elements

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js (hunk 2 above)
Comment (oscar-richardson-softwire) [reply e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:71]: Created cypress commands and updated row assertion to check deal ID column specifically

File: portal/component-tests/_macros/pagination.component-test.js [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+describe(component, () => {
+  let wrapper;
+  let totalPages;
+  let currentPage;
+  const totalItems = 2000;
+  const paginationRoute = '/testRoute';
+  const queryString = '?testQuery=test';
+
+  const shouldRenderOuterDivAndDisplayTotalItems = () => {
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply portal/component-tests/_macros/pagination.component-test.js:15]: Changed to functions and moved to the bottom of the describe block

File: portal/component-tests/_macros/pagination.component-test.js (hunk 3 above)
Comment (oscar-richardson-softwire) [reply portal/component-tests/_macros/pagination.component-test.js:136]: Personally, I think it makes sense for these functions to remain in this test file - they are specific to this pagination component and won't be used anywhere else.

Just to clarify, RE your suggestion about the `data-cy` selectors, do you mean replacing e.g., `[data-cy="Page_${i}_listItem"]` with a function that returns this, e.g., listItemForPage(i)?

File: trade-finance-manager-ui/component-tests/_macros/pagination.component-test.js [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+
+describe(component, () => {
+  let wrapper;
+  let totalPages;
+  let currentPage;
+  const totalItems = 2000;
+  const paginationRoute = '/testRoute';
+  const queryString = '?testQuery=test';
+
+  const shouldRenderOuterDivAndDisplayTotalItems = () => {
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/component-tests/_macros/pagination.component-test.js:15]: Changed to functions and moved to bottom of describe block

File: trade-finance-manager-ui/server/api.test.js (hunk 5 above)
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/server/api.test.js:118]: Great spot, I've done this :)

File: trade-finance-manager-ui/server/controllers/deals/index.test.js [hunk 10]
```diff
@@ -1,48 +1,266 @@
...
-        expect(res.render).toHaveBeenCalledWith('deals/deals.njk', {
-          heading: 'All deals',
-          deals: mockDeals,
-          activePrimaryNavigation: 'all deals',
-          activeSubNavigation: 'deal',
-          sortButtonWasClicked: false,
-          user: mockReq.session.user,
-          activeSortByField: 'tfm.dateReceivedTimestamp',
-          activeSortByOrder: 'descending',
+        shouldMakeRequestsForDealsDataWithDefaultArguments(mockReq);
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/server/controllers/deals/index.test.js:121]: Changed (and also changed in the other places I've created similarly named functions)

File: trade-finance-manager-ui/server/routes/deals/index.test.js [hunk 11]
```diff
@@ -12,10 +12,12 @@ describe('routes - deals', () => {
   });
 
   it('should setup routes with controllers', () => {
-    expect(get).toHaveBeenCalledTimes(1);
-    expect(post).toHaveBeenCalledTimes(1);
+    expect(get).toHaveBeenCalledTimes(2);
+    expect(post).toHaveBeenCalledTimes(2);
 
-    expect(get).toHaveBeenCalledWith('/', getDeals);
+    expect(get).toHaveBeenCalledWith('/', expect.any(Function));
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/server/routes/deals/index.test.js:18]: It's called with the anonymous function `(req, res) => res.redirect('/deals/0')`, but I couldn't find a good way of testing this except for naming the function and exporting it, but exporting just for the sake of testing felt like bad practice

File: trade-finance-manager-ui/server/controllers/deals/index.js [hunk 12]
```diff
@@ -2,79 +2,66 @@ const api = require('../../api');
...
-  const queryParams = {
-    sortBy: CONSTANTS.DEALS.TFM_SORT_BY_DEFAULT,
-  };
-  const { userToken } = req.session;
-
-  const apiResponse = await api.getDeals(queryParams, userToken);
-  const { data: amendments } = await api.getAllAmendmentsInProgress(userToken);
-
-  // override the deal stage if there is an amendment in progress
-  if (Array.isArray(amendments) && amendments?.length > 0) {
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/server/controllers/deals/index.js:15]: It is needed - it looks like this comment is on an old version of the code. I think I was testing to see if it was necessary by getting rid of it and committed that change by mistake

File: portal/component-tests/_macros/pagination.component-test.js [hunk 13]
```diff
@@ -0,0 +1,637 @@
...
+
+const render = componentRenderer(component);
+
+describe(component, () => {
+  let wrapper;
+  const totalItems = 2000;
+  const paginationRoute = '/testRoute';
+  const queryString = '?testQuery=test';
+
+  describe('when totalPages is small enough that all page links can be rendered on all pages (and an ellipsis is never rendered)', () => {
```
Comment (AlexBramhill) [inline portal/component-tests/_macros/pagination.component-test.js:13]: Does this have to be so specific? It probably does as it is very specific wording. If not:

```suggestion
  describe('when totalPages is small enough that an ellipsis is never rendered', () => {
```

File: portal/component-tests/_macros/pagination.component-test.js [hunk 14]
```diff
@@ -0,0 +1,637 @@
...
+          currentPage,
+          totalItems,
+          paginationRoute,
+          queryString,
+        });
+      });
+
+      itShouldRenderOuterDivAndDisplayTotalItems();
+
+      itShouldNOTRenderNavigationList();
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply portal/component-tests/_macros/pagination.component-test.js]: Changed to `Not`

**Review by AlexBramhill — APPROVED** [review]
lgtm

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -8,7 +8,7 @@ run-name: Executing test QA on ${{ github.repository }} 🚀
 
 on:
   pull_request:
-    branches: [main]
+    branches: [main, feat/dtfs2-6113/add-pagination-to-tfm-deals]
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply .github/workflows/test.yml:11]: Bump 😛

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  });
+
+  it('should allow the user to navigate to the previous page of the deals table', () => {
+    cy.visit('/deals/3');
+    pages.dealsPage.pagination.previous().click();
+
+    cy.url().should('eq', relative('/deals/2'));
+    cy.checkDealIdCell({ rowNumber: 0, expectedDealId: '10000041' });
+    cy.checkDealIdCell({ rowNumber: 19, expectedDealId: '10000060' });
+    cy.checkDealsTableRowsTotal(20);
```
Comment (ttbarnes) [inline e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:52]: I'm wondering if we need to check each row/deal ID rather than 2 of 20, maybe could have something like

```js
cy.checkDealsTableRows({ dealIds, totalRows = 20 });
```

And inside this command can call `checkDealIdCell()` X times and increment from a single `dealId` or passed in deal IDs. What do you think?

Thinking this could help to ensure initial sorting/ordering is working as expected? Not sure if this is handled somewhere else

File: e2e-tests/tfm/cypress/e2e/pages/dealsPage.js [hunk 17]
```diff
@@ -3,6 +3,7 @@ const dealsPage = {
   searchFormInput: () => cy.get('[data-cy="search-input"]'),
   searchFormSubmitButton: () => cy.get('[data-cy="submit-button"]'),
   dealsTableRows: () => cy.get('[data-cy="deals-table"] tbody tr'),
+  dealIdCell: (rowNumber) => cy.get('[data-cy="deals-table"] tbody tr').eq(rowNumber).children('td').eq(0),
```
Comment (ttbarnes) [inline e2e-tests/tfm/cypress/e2e/pages/dealsPage.js:6]: Should this be in`dealsTable.row` below? Looks like other body cells/columns are in there

File: trade-finance-manager-ui/server/controllers/deals/index.test.js [hunk 18]
```diff
@@ -55,204 +220,195 @@ describe('controllers - deals', () => {
...
+          pagesize: 20,
+          page: 0,
+        },
+        'userToken',
+      );
+      expect(api.getAllAmendmentsInProgress).toHaveBeenCalledWith('userToken');
+    });
+  }
+
+  function itShouldRenderDealsTemplateWithDefaultArguments({ mockReq, overrideDealStage }) {
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/deals/index.test.js:380]: What's the reason for wrapping the tests in these functions? Doesn't look like they're used in other places, wondering if they could just be `it('...`?

File: portal/component-tests/_macros/pagination.component-test.js (hunk 3 above)
Comment (oscar-richardson-softwire) [reply portal/component-tests/_macros/pagination.component-test.js]: I've moved the functions into their own directory `test-assertions`.

As discussed on Slack, I've not replaced the `data-cy` selectors with constants, although this could be part of a future refactor

File: trade-finance-manager-ui/server/controllers/deals/index.test.js (hunk 18 above)
Comment (oscar-richardson-softwire) [reply trade-finance-manager-ui/server/controllers/deals/index.test.js:380]: The reason is simply to make the tests in this file more DRY. This (long) function is called in 3 places in this file

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js (hunk 16 above)
Comment (oscar-richardson-softwire) [reply e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:48]: Good shout, I've replaced with a command of this form:
`cy.checkDealIdCells({ firstDealId: '10000001', increment: 1, numberToCheck: 20 });`

File: portal/component-tests/test-assertions/pagination-assertions.js [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+
+  it('should NOT render a \'Last\' link', () => {
+    const wrapper = getWrapper();
+
+    wrapper.expectElement('[data-cy="Last_listItem"]').notToExist();
+
+    wrapper.expectLink('[data-cy="Last"]').notToExist();
+  });
+}
+
```
Comment (ttbarnes) [inline portal/component-tests/test-assertions/pagination-assertions.js:197]: Nice, maybe could rename the file so it's `test-assertions/pagination`? Since the parent directory already has "assertions"

What do you think about splitting these up more so each function is in it's own file in a `pagination` directory, then this file can be a simple `index` with imports/exports?

Comment (oscar-richardson-softwire) [reply portal/component-tests/test-assertions/pagination-assertions.js]: `What do you think about splitting these up more so each function is in it's own file in a pagination directory, then this file can be a simple index with imports/exports?`

I've done this :)

File: e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+
+    pages.dealsPage.pagination.next().click();
+
+    pages.dealsPage.heading().invoke('text').then((text) => {
+      expect(text.trim()).to.equal(`${expectedNumberOfMatches} results for "${searchString}"`);
+    });
+    cy.checkDealIdCells({ firstDealId: '10000041', increment: 2, numberToCheck: 10 });
+    cy.checkDealsTableRowsTotal(11);
+  });
+
```
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply e2e-tests/tfm/cypress/e2e/journeys/deals/deals-pagination.spec.js:105]: I agree that this should be done as tech debt along with other E2E tests

File: .github/workflows/test.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (oscar-richardson-softwire) [reply .github/workflows/test.yml:11]: Removed
